File: LeetCode/LeetCode Practice Problems/1200- Minimum Absolute Difference.cpp
class Solution {
public:
    vector<vector<int>> minimumAbsDifference(vector<int>& arr) {
        vector<vector<int>> ans;
        sort(arr.begin() , arr.end());
        int n = arr.size();
        
        int mini = INT_MAX;
        
        for(int i=0;i<n-1;i++)
        {
            int v = arr[i+1] - arr[i];
            if(mini > v)
            {
                ans.clear();
                mini = v;
            }
            if(mini == v)
            {
                ans.push_back({arr[i] , arr[i+1]});
            }
        }
        
        
        return ans;
    }
};

File: LeetCode/LeetCode Practice Problems/1200-Minimum Absolute Difference.java
class Solution {
    public List<List<Integer>> minimumAbsDifference(int[] arr) {
        int minElement = arr[0];
        int maxElement = arr[0];
        
        for(int num: arr) {
            minElement = Math.min(minElement , num);
            maxElement = Math.max(maxElement , num);
        }
        
        int shift = -minElement;
        int cnt[] = new int[maxElement - minElement + 1];
        
        for(int num: arr) {
            cnt[num + shift] = 1;
        }
        
        int minPairDiff = maxElement - minElement;
        int prev = 0;
        List<List<Integer>> ans = new ArrayList();
        
        for(int cur=1;cur<=maxElement+shift;cur++) {
            if(cnt[cur] == 0) continue;
            
            if(cur - prev < minPairDiff) {
                minPairDiff = cur - prev;
                ans.clear();
            }
            if(cur - prev == minPairDiff) {
                ans.add(Arrays.asList(prev - shift , cur - shift));
            }
            prev = cur;
        }
        
        return ans;
    }
}
